### ISSUE TITLE
GMT 0009-2012 User ID default string

### ISSUE THREAD

**Issue by ronaldtse** [riboseinc/rfc-openpgp-sca#13]
The default string for user identification is a 16 byte sequence:
```
0x31,0x32,0x33,0x34,0x35,0x36,0x37,0x38,0x31,0x32,0x33,0x34,0x35,0x36,0x37,0x38
```

**Comment by ronaldtse** [comment]
@randombit we're using the empty string "``" right now. There's probably no point to change to this "1234567812345678" here, right?

**Comment by ronaldtse** [comment]
Thoughts on this @dewyatt?

**Comment by dewyatt** [comment]
I believe the "SM2 user identity" is (correctly) a blank string after https://github.com/riboseinc/rnp/commit/c751be3c555419cbc87e011a0a679c44c6ea2afa in rnp.

Or is this referring to something else? My memory is foggy on this one.

**Comment by ronaldtse** [comment]
Yes, it is currently a blank string. However, GM/T 0009 specifies that the unknown identity for SM2 should use the crafty string "1234567812345678". Do you think a "blank string" is a better choice?

**Comment by dewyatt** [comment]
Oh I see what you're saying. I was saying it looked correct according to the text of our RFC here.

> SM2DSA allows use of an optional "user identity" string which is hashed into ZA (Section 3.5 of [SM2-2] and Section 5.1.4.4 of [I-D.shen-sm2-ecdsa]). In OpenPGP, the user identifier IDA MUST be the empty string.

I have not read this GMT document, so I can't really provide any input on whether we ought to switch to "1234567812345678". If it's clearly stated as a requirement, it seems like it wouldn't hurt to change it?

**Comment by ronaldtse** [comment]
Here's the [GM/T 0009-2012 standard](https://riboseinc.github.io/gbt-0009-2012/) I put up earlier on Monday.

**Comment by dewyatt** [comment]
It sort of sounds like botan might want to switch to that as the default here (looks like it would need some small code changes):
https://github.com/randombit/botan/blob/27482d71e00dcf106735ef824ded70cf25c6150f/src/lib/pubkey/sm2/sm2.cpp#L230
and here:
https://github.com/randombit/botan/blob/27482d71e00dcf106735ef824ded70cf25c6150f/src/lib/pubkey/sm2/sm2.cpp#L205

> There's probably no point to change to this "1234567812345678" here, right?

I wonder if it would make botan more compatible with other SM2 implementations?

(I also have to wonder if a blank userid reduces security of SM2 signatures at all?)

**Comment by ronaldtse** [comment]
Actually you're right: [GmSSL does use it as the default User ID](https://github.com/guanzhi/GmSSL/search?utf8=✓&q=SM2_DEFAULT_ID_GMT09&type=).

We should make a PR for botan for this issue -- and probably in OpenPGP where we use SM2 for public key encryption when the user id is not specified.

Could you help make this change? Thanks!

cc: @randombit

**Comment by dewyatt** [comment]
@ronaldtse I created the PR for botan.

> We should make a PR [...] in OpenPGP where we use SM2 for public key encryption when the user id is not specified.

What did you mean here? I think the SM2 ID is only used for signatures, right? (and is essentially incompatible with the OpenPGP userid and mostly irrelevant for us)

> https://riboseinc.github.io/gmt-0009-2012/

Minor detail, but did you intend for the URL to be "gbt" rather than "gmt"?

**Comment by ronaldtse** [comment]
I got confused. Indeed the ID is only used in Signatures and Key Exchange. And great catch -- it was a typo indeed... will fix it. Thanks for making the PR.
